build-system/eslint-rules/enums.js

```diff
@@ -0,0 +1,195 @@
+'use strict';
+
+/**
+ * Enforces all enum usage is staticly DCE-able.
+ *
+ * This is an opt-in lint, with enums enabling by using a "_Enum" suffix.
+ *
+ * Good
+ * ```
+ * /** @enum *\/
+ * const Foo_Enum = {
+ *   ABC: 1,
+ *   DEF: 2,
+ * };
+ *
+ * const { ABC: value } = Foo_Enum;
+ * const value = Foo_Enum.ABC;
+ * isEnumValue(Foo_Enum, value);
+ * ```
+ *
+ * Bad
+ * ```
+ * const Bar_Enum = {
+ *   0: 0,
+ *   '1': 1,
+ *   ['2'] : 2,
+ *   [ref]: 4,
+ * };
+ *
+ * Bar_Enum[0];
+ * Bar_Enum['0'];
+ * Bar_Enum['key'];
+ * Bar_Enum[ref];
+ * key in Bar_Enum;
+ * ```
+ *
+ * @return {!Object}
+ */
+module.exports = function (context) {
+  /**
+   * @param {!Node} node
+   * @return {boolean}
+   */
+  function hasEnumAnnotation(node) {
+    const commentLines = context.getCommentsBefore(node);
+    if (!commentLines) {
+      return false;
+    }
+
+    return commentLines.some(function (comment) {
+      return comment.type == 'Block' && /@enum/.test(comment.value);
+    });
+  }
+
+  /**
+   * @param {!Node|undefined} node
+   */
+  function checkEnumId(node) {
+    if (/^[A-Z](?:[A-Za-z0-9]+_Enum|[A-Z0-9_]+_ENUM)$/.test(node.name)) {
+      return;
+    }
+    context.report({
+      node,
+      message:
+        'Enums should use PascalCase and end in "_Enum", or SCREAMING_SNAKE_CASE and end in "_ENUM"',
+    });
+  }
+
+  /**
+   * @param {!Node|undefined} node
+   */
+  function checkEnumKeys(node) {
+    for (const prop of node.properties) {
+      if (prop.computed || prop.key?.type !== 'Identifier') {
+        context.report({
+          node: prop,
+          message: [
+            'Enum keys must be a normal prop identifier',
+            'eg, `{ KEY: value }` ',
+          ].join('\n\t'),
+        });
+      }
+    }
+  }
+
+  /**
+   * @param {!Node} node
+   */
+  function checkStaticEnumUse(node) {
+    const {parent} = node;
+    // import { Enum } from '.'
```

Comment:
```suggestion
```

Reply:
The comments are illustrative of what we're matching.